### PR TITLE
Implemented FilBlockTracker and added Filecoin.StateSearchMsgLimited and Filecoin.ChainHead requests

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/22099

Implemented `FilBlockTracker` to watch Filecoin requests via `Filecoin.StateSearchMsgLimited` and `Filecoin.ChainHead`
## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

-  no manual tests required

### CODE REVIEW

File: components/brave_wallet/browser/fil_tx_manager.cc [hunk 1]
```diff
@@ -224,6 +228,48 @@ FilTxStateManager* FilTxManager::GetFilTxStateManager() {
...
 
-void FilTxManager::UpdatePendingTransactions() {}
+void FilTxManager::UpdatePendingTransactions() {
+  auto pending_transactions = tx_state_manager_->GetTransactionsByStatus(
+      mojom::TransactionStatus::Submitted, absl::nullopt);
+  for (const auto& pending_transaction : pending_transactions) {
+    auto cid = pending_transaction->tx_hash();
+    uint64_t seconds =
+        (base::Time::Now() - pending_transaction->submitted_time()).InSeconds();
+    uint64_t depth = seconds == 0 ? 1 : seconds;
```
Comment (yrliou) [inline components/brave_wallet/browser/fil_tx_manager.cc:238]: I'm not sure I understand what's the logic of depth here, could you point me to some documentation of this?

File: components/brave_wallet/browser/fil_tx_manager.cc [hunk 2]
```diff
@@ -224,6 +228,48 @@ FilTxStateManager* FilTxManager::GetFilTxStateManager() {
...
+    json_rpc_service_->GetFilStateSearchMsgLimited(
+        cid, depth,
+        base::BindOnce(&FilTxManager::OnGetTransactionStatus,
```
Comment (yrliou) [inline components/brave_wallet/browser/fil_tx_manager.cc:241]: Should probably just use `OnGetFilStateSearchMsgLimited`.

File: components/brave_wallet/browser/fil_tx_manager.cc [hunk 3]
```diff
@@ -224,6 +228,48 @@ FilTxStateManager* FilTxManager::GetFilTxStateManager() {
...
+                       weak_factory_.GetWeakPtr(), pending_transaction->id()));
+  }
+  known_no_pending_tx_ = pending_transactions.empty();
+  CheckIfBlockTrackerShouldRun();
+}
+
+void FilTxManager::OnGetTransactionStatus(const std::string& tx_meta_id,
+                                          int exit_code,
```
Comment (yrliou) [inline components/brave_wallet/browser/fil_tx_manager.cc:249]: shouldn't it be int64?

File: components/brave_wallet/browser/fil_block_tracker.cc [hunk 4]
```diff
@@ -5,15 +5,78 @@
...
+               base::BindRepeating(&FilBlockTracker::GetLatestBlockhash,
+                                   weak_ptr_factory_.GetWeakPtr(),
+                                   base::NullCallback(), false));
+}
+
+void FilBlockTracker::Stop() {
+  BlockTracker::Stop();
+}
+
+void FilBlockTracker::GetLatestBlockhash(GetLatestBlockhashCallback callback,
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_block_tracker.cc:38]: yes, the Filecoin has also returns a hash of mined block, I can rename it to CID if need, thought we added a cache to return cached value if we have multiple calls at same time, want me to remove it?

File: components/brave_wallet/browser/fil_tx_manager.cc (hunk 1 above)
Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc:238]: `StateSearchMsgLimited` looks back up to limit epochs in the chain for a message, and returns its receipt and the tipset where it was executed. There is no documentation to clarify this parameter unfortunately, guys from Filecoin slack advised to pass amount of blocks based on time needed for a block to be mined, they said it usually 10-30 seconds, but experimentally I found that looking through 1 block per second work more stable.

File: components/brave_wallet/browser/json_rpc_service_unittest.cc [hunk 5]
```diff
@@ -2992,4 +3012,61 @@ TEST_F(JsonRpcServiceUnitTest, GetFilChainHead) {
...
+        "result":
+        {
+            "Height": 22389,
+            "Message":
+            {
+                "/": "bafy2bzacebundyopm3trenj47hxkwiqn2cbvvftz3fss4dxuttu2u6xbbtkqy"
+            },
+            "Receipt":
+            {
+                "ExitCode": 0,
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc:3031]: zero means the transaction was successful,  I need it for this test

File: components/brave_wallet/browser/fil_response_parser.cc [hunk 6]
```diff
@@ -73,4 +73,26 @@ bool ParseFilGetChainHead(const std::string& json, std::string* cid) {
   return true;
 }
 
+// Returns parsed receipt exit code.
+bool ParseFilStateSearchMsgLimited(const std::string& raw_json,
+                                   const std::string& cid,
+                                   int64_t* exit_code) {
+  std::string converted_json(json::convert_int64_value_to_string(
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_response_parser.cc]: moved, thanks

File: components/brave_wallet/browser/json_rpc_service_unittest.cc [hunk 7]
```diff
@@ -576,6 +576,23 @@ class JsonRpcServiceUnitTest : public testing::Test {
         }));
     run_loop.Run();
   }
+  void GetFilChainHead(const std::string& expected_cid,
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc:611]: done

File: components/brave_wallet/browser/fil_block_tracker.cc [hunk 8]
```diff
@@ -5,15 +5,78 @@
...
-  NOTIMPLEMENTED();
+  timer_.Start(FROM_HERE, interval,
+               base::BindRepeating(&FilBlockTracker::GetLatestBlockhash,
+                                   weak_ptr_factory_.GetWeakPtr(),
+                                   base::NullCallback(), false));
+}
+
+void FilBlockTracker::Stop() {
+  BlockTracker::Stop();
+}
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_block_tracker.cc]: done

File: components/brave_wallet/browser/fil_tx_manager.cc (hunk 2 above)
Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/browser/fil_tx_manager.cc (hunk 3 above)
Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/browser/json_rpc_service.cc [hunk 9]
```diff
@@ -675,6 +675,45 @@ void JsonRpcService::OnFilGetBalance(
...
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetFilChainHead,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  RequestInternal(fil::getChainHead(), true, network_url,
+                  std::move(internal_callback));
+}
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/json_rpc_service.cc:738]: done

File: components/brave_wallet/browser/fil_requests.cc [hunk 10]
```diff
@@ -74,6 +74,17 @@ std::string getChainHead() {
...
+  base::Value cid_value(base::Value::Type::DICTIONARY);
+  cid_value.SetStringKey("/", cid);
+  auto result =
+      GetJsonRpc2Params("Filecoin.StateSearchMsgLimited", std::move(cid_value),
+                        base::Value(std::to_string(period)));
+  result = std::string(
+      json::convert_string_value_to_uint64("/params/1", result.c_str(), false)
+          .c_str());
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_requests.cc:87]: done

File: components/brave_wallet/browser/json_rpc_service_unittest.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/json_rpc_service_unittest.cc]: wrapped it to a function and added MAX test case as sell

File: components/brave_wallet/browser/fil_response_parser.cc [hunk 11]
```diff
@@ -55,4 +55,22 @@ bool ParseFilEstimateGas(const std::string& json,
   return true;
 }
 
+bool ParseFilGetChainHead(const std::string& json, std::string* cid) {
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_response_parser.cc:58]: There is no other documentation, unfortunately. These api are used to check if the head has changed. It is enough to check the head cid or Height. I can change it to Height if it looks more understandable.

File: components/brave_wallet/browser/fil_block_tracker.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/fil_block_tracker.cc:34]: I prefer not to add the complexity without a real use case for now, thanks.

File: components/brave_wallet/browser/fil_response_parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/fil_response_parser.cc:58]: Yea, height is more understandable, thanks.

File: components/brave_wallet/browser/fil_tx_manager.cc (hunk 1 above)
Comment (yrliou) [reply components/brave_wallet/browser/fil_tx_manager.cc:238]: Please document this in the code, thanks.
Also, I think limit_epochs is more suitable than depth.

File: components/brave_wallet/browser/fil_block_tracker.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_block_tracker.cc]: done

File: components/brave_wallet/browser/fil_tx_manager.cc (hunk 1 above)
Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/browser/fil_response_parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/fil_response_parser.cc]: done

File: components/brave_wallet/browser/fil_response_parser_unittest.cc [hunk 12]
```diff
@@ -180,4 +188,79 @@ TEST(FilResponseParserUnitTest, ParseFilEstimateGas) {
                                                  &gas_fee_cap, &gas_limit));
 }
 
+TEST(FilResponseParserUnitTest, ParseFilGetChainHead) {
```
Comment (yrliou) [inline components/brave_wallet/browser/fil_response_parser_unittest.cc:191]: Can we add a case for height being an empty string?

File: components/brave_wallet/browser/fil_tx_manager.cc [hunk 13]
```diff
@@ -224,6 +228,53 @@ FilTxStateManager* FilTxManager::GetFilTxStateManager() {
...
+    return;
+  std::unique_ptr<FilTxMeta> meta =
+      GetFilTxStateManager()->GetFilTx(tx_meta_id);
+  if (!meta)
+    return;
+  mojom::TransactionStatus status = (exit_code == 0)
+                                        ? mojom::TransactionStatus::Confirmed
+                                        : mojom::TransactionStatus::Error;
+  meta->set_status(status);
+  meta->set_confirmed_time(base::Time::Now());
```
Comment (yrliou) [inline components/brave_wallet/browser/fil_tx_manager.cc:268]: I think currently for other coin types, we only do `set_confirmed_time` if it's confirmed and not error.

Comment (spylogsster) [reply components/brave_wallet/browser/fil_tx_manager.cc]: done

File: components/brave_wallet/browser/fil_response_parser_unittest.cc (hunk 12 above)
Comment (spylogsster) [reply components/brave_wallet/browser/fil_response_parser_unittest.cc:191]: done

**Review by yrliou — APPROVED** [review]
LGTM

**Pull request merged**